File: go/nato.go
// go run nato.go

package main

import (
	"encoding/json"
	"fmt"
	"log"
)

func main() {
	handle_json()
}

// See ../1-json-primer.md for an explanation
func handle_json() {
	jsonText := `{"foo": {"bar": [{"nato": "oscar"}, {"nato": "papa"}, {"nato": "quebec"}]}}`
	data := make(map[string]interface{})
	err := json.Unmarshal([]byte(jsonText), &data)
	if err != nil {
		log.Fatal(err)
	}

	// data["foo"]["bar"][1]["nato"]
	natoLetter := data["foo"].
		(map[string]interface{})["bar"].
		([]interface{})[1].
		(map[string]interface{})["nato"]  // "papa"
	fmt.Printf("%v\n", natoLetter)

	// data["foo"]["quux"] = {"stuff": "nonsense", "nums": [2.718, 3.142]}
	quux := map[string]interface{}{
		"stuff": "nonsense",
		"nums":  []float64{2.718, 3.142},
	}
	data["foo"].(map[string]interface{})["quux"] = quux

	result, err := json.MarshalIndent(data, "", "    ")
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("%s\n", result)
}

File: csharp/nato.cs
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Demo
{
    class Program
    {
        static void Main(string[] args)
        {
            handle_json();
        }

        // See ../1-json-primer.md for an explanation
        static void handle_json()
        {
            string json_text = "{\"foo\": {\"bar\": [{\"nato\": \"oscar\"}, {\"nato\": \"papa\"}, {\"nato\": \"quebec\"}]}}";
            dynamic data = JsonConvert.DeserializeObject(json_text);

            string nato_letter = data.foo.bar[1].nato;  // "papa"
            Console.WriteLine(nato_letter);

            // data["foo"]["quux"] = {"stuff": "nonsense", "nums": [2.718, 3.142]}
            data.foo.quux = JObject.FromObject(
                new {
                    stuff = "nonsense",
                    nums = new[] {2.718, 3.142}
                }
            );

            string result = JsonConvert.SerializeObject(data, Formatting.Indented);
            Console.WriteLine(result);
        }
    }
}
